### PR TITLE
New method to revert emoji String

### PR DESCRIPTION
**Pull request by pepicrft**
I've created a new method that passing a emojized string return the one with emojis reverted into original chearsheet value.

### CODE REVIEW

File: NSStringEmojize/NSString+Emojize.m [hunk 1]
```diff
@@ -45,7 +45,18 @@ + (NSString *)emojizedStringWithString:(NSString *)text
...
+    static NSRegularExpression *regex = nil;
+    dispatch_once(&onceToken, ^{
+        regex = [[NSRegularExpression alloc] initWithPattern:@"(\\UD000[A-F0-9]{5})" options:NSRegularExpressionCaseInsensitive error:NULL];
+    });
+    NSString *resultText = text;
+    for (NSString *key in [self.emojiAliases allKeys]){
+        resultText = [resultText stringByReplacingOccurrencesOfString:self.emojiAliases[key]  withString:key];
+    }
+    return resultText;
+}
```
Comment (thisandagain) [inline NSStringEmojize/NSString+Emojize.m:59]: Looks great... just need a line break here.

Comment (pepicrft) [inline NSStringEmojize/NSString+Emojize.m:59]: Where exactly @thisandagain ?

**Comment by ghost** [comment]
I think @thisandagain was suggesting a line break was needed before and after the method implementation to be consistent with the style in the rest of the file.

One strange bit is the `NSRegularExpression`. While declared and initialized, it is unused and instead the source string is tested against all 844 aliases. What are the performance implications for manually testing vs using an `NSRegularExpression`? :weary:

**Comment by ghost** [comment]
Another thing worth noting is that there are currently **845 unique keys** but only **821 unique values**, so mapping from value to key will be imperfect and likely to produce some oddities like `:-1:` but no matching `:+1:` with instead `:thumbsup:`. So a bias list of some sort might be needed to provide more self-consistent results, no?

**Comment by ghost** [comment]
And the overlaps are in :collision: :poop: :laughing: :moon: :book: :memo: :phone: :train: :red_car: :running: :bee: :dolphin: :lantern: :paw_prints: :punch: :sailboat: :+1: :-1: :shirt: :email: :hand: :shoe: :heavy_exclamation_mark:

via:

``` objective-c
NSCountedSet* emojiSet = [NSCountedSet setWithArray:[emojiAliases() allValues]];
for (NSString* e in emojiSet) {
    if ([emojiSet countForObject:e] == 1) {
        continue;
    }

    NSSet* keys = [emojiAliases() keysOfEntriesPassingTest:^BOOL(NSString* key, NSString* obj, BOOL* stop) {
        return [obj isEqualToString:e];
    }];
    NSLog(@"%@", keys);
}
```

which outputs:

```
2014-01-03 18:15:24.878 Generator[7913:303] {(
    ":collision:",
    ":boom:"
)}
2014-01-03 18:15:24.879 Generator[7913:303] {(
    ":shit:",
    ":poop:",
    ":hankey:"
)}
2014-01-03 18:15:24.880 Generator[7913:303] {(
    ":laughing:",
    ":satisfied:"
)}
2014-01-03 18:15:24.880 Generator[7913:303] {(
    ":moon:",
    ":waxing_gibbous_moon:"
)}
2014-01-03 18:15:24.881 Generator[7913:303] {(
    ":book:",
    ":open_book:"
)}
2014-01-03 18:15:24.881 Generator[7913:303] {(
    ":memo:",
    ":pencil:"
)}
2014-01-03 18:15:24.882 Generator[7913:303] {(
    ":phone:",
    ":telephone:"
)}
2014-01-03 18:15:24.882 Generator[7913:303] {(
    ":train:",
    ":railway_car:"
)}
2014-01-03 18:15:24.883 Generator[7913:303] {(
    ":red_car:",
    ":car:"
)}
2014-01-03 18:15:24.883 Generator[7913:303] {(
    ":running:",
    ":runner:"
)}
2014-01-03 18:15:24.884 Generator[7913:303] {(
    ":bee:",
    ":honeybee:"
)}
2014-01-03 18:15:24.884 Generator[7913:303] {(
    ":dolphin:",
    ":flipper:"
)}
2014-01-03 18:15:24.885 Generator[7913:303] {(
    ":lantern:",
    ":izakaya_lantern:"
)}
2014-01-03 18:15:24.885 Generator[7913:303] {(
    ":paw_prints:",
    ":feet:"
)}
2014-01-03 18:15:24.886 Generator[7913:303] {(
    ":punch:",
    ":facepunch:"
)}
2014-01-03 18:15:24.886 Generator[7913:303] {(
    ":sailboat:",
    ":boat:"
)}
2014-01-03 18:15:24.886 Generator[7913:303] {(
    ":+1:",
    ":thumbsup:"
)}
2014-01-03 18:15:24.887 Generator[7913:303] {(
    ":-1:",
    ":thumbsdown:"
)}
2014-01-03 18:15:24.887 Generator[7913:303] {(
    ":shirt:",
    ":tshirt:"
)}
2014-01-03 18:15:24.887 Generator[7913:303] {(
    ":email:",
    ":envelope:"
)}
2014-01-03 18:15:24.888 Generator[7913:303] {(
    ":hand:",
    ":raised_hand:"
)}
2014-01-03 18:15:24.888 Generator[7913:303] {(
    ":shoe:",
    ":mans_shoe:"
)}
2014-01-03 18:15:24.889 Generator[7913:303] {(
    ":heavy_exclamation_mark:",
    ":exclamation:"
)}
```

**Comment by thisandagain** [comment]
I wasn't too worried about the overlaps, but I'm wondering if we should write a script to build a reverse lookup dictionary rather than iterate across the entire contents of the existing dictionary. The performance costs of the approach you have here can be high.

**Comment by thisandagain** [comment]
I'm going to close this for now and add a feature request where we can track this.

**Comment by ghost** [comment]
Nah, you don't need a script, just do it at runtime wrapped in a `dispatch_once` much like the original `NSDictionary`, the code above was just to demonstrate the overlaps. An `NSRegularExpression` instead of `-[NSString stringByReplacingOccurrencesOfString:]` would definitely be beneficial too!

**Pull request closed without merging**